subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap.c

```diff
@@ -0,0 +1,489 @@
+		result_code / 32u, result_code & 0x1f, offset, len, last_block);
+	if (result_code != COAP_RESPONSE_CODE_CONTENT) {
+		fota_err = result_code;
+	} else if (payload && len) {
+		LOG_INF("Got FOTA response: %.*s", len, (const char *)payload);
+		fota_err = coap_codec_fota_resp_decode(user, payload, len,
+						       COAP_CONTENT_FORMAT_APP_JSON);
+	} else {
+		fota_err = 0;
+	}
+}
+
+int nrf_cloud_coap_fota_job_get(struct nrf_cloud_fota_job_info *const job)
+{
+	__ASSERT_NO_MSG(job != NULL);
+	CHECK_AUTH();
+	int err;
+
+	job->type = NRF_CLOUD_FOTA_TYPE__INVALID;
+
+	err = nrf_cloud_coap_get("fota/exec/current", NULL, NULL, 0,
+				 COAP_CONTENT_FORMAT_APP_CBOR,
+				 COAP_CONTENT_FORMAT_APP_JSON, true, get_fota_callback, job);
+
+	if (!err && !fota_err) {
+		LOG_INF("FOTA job received; type:%d, id:%s, host:%s, path:%s, size:%d",
+			job->type, job->id, job->host, job->path, job->file_size);
+	} else if (!err && (fota_err == COAP_RESPONSE_CODE_NOT_FOUND)) {
+		LOG_INF("No pending FOTA job");
+	} else if (err == -EAGAIN) {
+		LOG_ERR("Timeout waiting for FOTA job");
+	} else {
+		LOG_ERR("Error getting current FOTA job; FOTA rc:%d.%02d, err:%d",
+			fota_err / 32u, fota_err & 0x1f, err);
+		err = fota_err;
+	}
+	return err;
+}
+
+void nrf_cloud_coap_fota_job_free(struct nrf_cloud_fota_job_info *const job)
+{
+	nrf_cloud_rest_fota_job_free(job);
+}
+
+int nrf_cloud_coap_fota_job_update(const char *const job_id,
+	const enum nrf_cloud_fota_status status, const char * const details)
+{
+	__ASSERT_NO_MSG(job_id != NULL);
+	__ASSERT_NO_MSG(status < JOB_STATUS_STRING_COUNT);
+	CHECK_AUTH();
+
+#define API_FOTA_JOB_EXEC		"fota/exec"
+#define API_UPDATE_FOTA_URL_TEMPLATE	(API_FOTA_JOB_EXEC "/%s")
+#define API_UPDATE_FOTA_BODY_TEMPLATE	"{\"status\":\"%s\"}"
+#define API_UPDATE_FOTA_DETAILS_TMPLT	"{\"status\":\"%s\", \"details\":\"%s\"}"
+	/* Mapping of enum to strings for Job Execution Status. */
+	static const char *const job_status_strings[] = {
+		[NRF_CLOUD_FOTA_QUEUED]      = "QUEUED",
+		[NRF_CLOUD_FOTA_IN_PROGRESS] = "IN_PROGRESS",
+		[NRF_CLOUD_FOTA_FAILED]      = "FAILED",
+		[NRF_CLOUD_FOTA_SUCCEEDED]   = "SUCCEEDED",
+		[NRF_CLOUD_FOTA_TIMED_OUT]   = "TIMED_OUT",
+		[NRF_CLOUD_FOTA_REJECTED]    = "REJECTED",
+		[NRF_CLOUD_FOTA_CANCELED]    = "CANCELLED",
+		[NRF_CLOUD_FOTA_DOWNLOADING] = "DOWNLOADING",
+	};
+#define JOB_STATUS_STRING_COUNT (sizeof(job_status_strings) / \
+				 sizeof(*job_status_strings))
+
+	int ret;
+	size_t buff_sz;
+	char *url = NULL;
+	char *payload = NULL;
+
+	/* Format API URL with device and job ID */
+	buff_sz = sizeof(API_UPDATE_FOTA_URL_TEMPLATE) + strlen(job_id);
+	url = nrf_cloud_malloc(buff_sz);
+	if (!url) {
+		ret = -ENOMEM;
+		goto clean_up;
+	}
+
+	ret = snprintk(url, buff_sz, API_UPDATE_FOTA_URL_TEMPLATE, job_id);
+	if ((ret < 0) || (ret >= buff_sz)) {
+		LOG_ERR("Could not format URL");
+		ret = -ETXTBSY;
+		goto clean_up;
+	}
+
+	/* Format payload */
+	if (details) {
```

Comment:
It would be better to abandon JSON (which is used here) and use CBOR.  Until then, if I were to move some of this into a shared function as you suggest in another comment, then the code still needs to handle details as well as no details.